cloudsync/providers/gdrive.py

```diff
@@ -133,17 +157,128 @@ def events(self, timeout):
+        mtime = arrow.get(mtime).isoformat()
+        gdrive_info = {
+                'modifiedTime':  mtime
+                }
+
+        # mime type, if provided
+        mime_type = metadata.get("mimeType", "application/octet-stream")
+        if mime_type:
+            gdrive_info['mimeType'] = mime_type
+
+        # path, if provided
+        if path:
+            parent, name = os.path.split(path)
+            gdrive_info['name'] = name
+
+        # misc properties, if provided
+        app_props = metadata.get("appProperties", None)
+        if app_props:
+            # caller can specify google-specific stuff, if desired
+            gdrive_info['appProperties'] = app_props
+
+        # misc properties, if provided
+        app_props = metadata.get("properties", None)
+        if app_props:
+            # caller can specify google-specific stuff, if desired
+            gdrive_info['properties'] = app_props
+
+        log.debug("info %s", gdrive_info)
+
+        return gdrive_info
+
+    def create(self, path, file_like, metadata={}) -> 'ProviderInfo':
+        gdrive_info = self.__prep_upload(path, file_like, metadata)
+
+        ul = MediaIoBaseUpload(file_like, mimetype=gdrive_info.get('mimeType'), chunksize=4 * 1024 * 1024)
+
+        fields = 'id, md5Checksum'
+
+        parent, name = os.path.split(path)
+        if not self.exists_path(parent):
+            raise CloudFileNotFoundError("parent folder %s doesn't exist", parent)
+        parent_oid = self._ids[parent]
+
+        gdrive_info['parents'] = [parent_oid]
+
+        res = self._api('files', 'create',
+                body=gdrive_info,
+                media_body=ul,
+                fields=fields)
+
+        log.debug("response from upload %s", res)
+
+        if not res:
+            raise CloudTemporaryError("unknown response from drive on upload")
+
+        self._ids[path] = res['id']
+
+        return ProviderInfo(oid=res['id'], hash=res['md5Checksum'], path=path)
 
     def download(self, oid, file_like):
-        ...
+        req = self.client.files().get_media(fileId=oid)
+        dl = MediaIoBaseDownload(file_like, req, chunksize=4 * 1024 * 1024)
+
+        done = False
+        while not done:
+            try:
+                status, done = dl.next_chunk()
+            except HttpError as e:
+                if str(e.resp.status) == '416':
+                    log.debug("empty file downloaded")
+                    done = True
+                elif str(e.resp.status) == '404':
+                    raise CloudFileNotFoundError("file %s not found", oid) 
+                else:
+                    raise CloudTemporaryError("unknown response from drive")
+            except (TimeoutError, HttpLib2Error) as e:
+                self.disconnect()
+                raise CloudDisconnectedError("disconnected during download")
 
     def rename(self, oid, path):
-        ...
+        parent, name = os.path.split(path)
+        
+        if not self.exists_path(parent):
```

Comment:
abstract this to a get_parent_pid function so that it is reusable and self-documenting